### PR TITLE
os/bluestore: a major refactor around allocmap persistency

### PR DESCRIPTION
**Pull request by ifed01**
This includes:
- introducing v.2 format for allocmap persistence to address 4G file limit
- adding straightforward control for choosing desired allocmap persistence mode during mkfs via bluestore_freelist_type parameter
- introducing a bunch of commands to ceph-bluestore-tool to change allocmap mode.
- refactoring code to have better test coverage
- adding a bunch of tests in store_test.cc

Fixes: https://tracker.ceph.com/issues/58646
Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/os/bluestore/simple_bitmap.cc [hunk 1]
```diff
@@ -274,3 +274,25 @@ extent_t SimpleBitmap::get_next_clr_extent(uint64_t offset)
   ext.length = (soffset - ext.offset);
   return ext;
 }
+//----------------------------------------------------------------------------
+int SimpleBitmap::compare(const SimpleBitmap& other) const
```
Comment (aclamk) [inline src/os/bluestore/simple_bitmap.cc:278]: It can be bool.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -19175,73 +19356,24 @@ int cmpfunc (const void * a, const void * b)
...
+
+  alloc1->foreach(
+    [&](uint64_t offset, uint64_t length) {
+      b1.set(offset >> order, length >> order);
+  });
+  alloc2->foreach(
+    [&](uint64_t offset, uint64_t length) {
+      b2.set(offset >> order, length >> order);
+  });
+  return 0;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:19376]: Missing actual compare between b1 and b2.

File: src/os/bluestore/Allocator.h [hunk 3]
```diff
@@ -55,6 +55,7 @@ class Allocator {
   virtual void dump() = 0;
   virtual void foreach(
     std::function<void(uint64_t offset, uint64_t length)> notify) = 0;
+  virtual const uint64_t* get_as_bitmap(size_t* out_count) const { return nullptr; }
```
Comment (aclamk) [inline src/os/bluestore/Allocator.h:58]: An open question:
is it a good practice to add a function to interface, in cases where we know that only one instance will be able to supply implementation.
Is not dynamic_cast better way to address this type of specific extensions?

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -18208,7 +18207,19 @@ bool RocksDBBlueFSVolumeSelector::compare(BlueFSVolumeSelector* other) {
 
 static const std::string allocator_dir    = "ALLOCATOR_NCB_DIR";
 static const std::string allocator_file   = "ALLOCATOR_NCB_FILE";
-static uint32_t    s_format_version = 0x01; // support future changes to allocator-map file
+static uint32_t    s_format_version = 0x02; // support future changes to allocator-map file
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:18210]: This is making mess.
Leave s_format_version = 0x01 as the offset-size version.
Create new bitmap_format_version = 0x02 as the new one.

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -18208,7 +18207,19 @@ bool RocksDBBlueFSVolumeSelector::compare(BlueFSVolumeSelector* other) {
...
+                                            // 0x00 - dummy
+                                            // 0x01 - original design, allocation
+                                            //        map is stored as a sequence
+                                            //        of <offset, length> pairs
+                                            // 0x02 - allocation map is storead
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:18215]: stored

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -18500,6 +18511,36 @@ int BlueStore::copy_allocator(Allocator* src_alloc, Allocator* dest_alloc, uint6
   return 0;
 }
 
+//-----------------------------------------------------------------------------------
+size_t calc_allocator_image_header_size()
+{
+  utime_t                 timestamp = ceph_clock_now();
+  allocator_image_header  header(timestamp, s_format_version, s_serial);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc]: I guess here will be a good place for "dummy_format_version"

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -18208,7 +18207,19 @@ bool RocksDBBlueFSVolumeSelector::compare(BlueFSVolumeSelector* other) {
...
+                                            //        of <offset, length> pairs
+                                            // 0x02 - allocation map is storead
+                                            //        as a bitmap.
+                                            // NOTE: we don't care about endianess
+                                            // for v2 format, in rare case one
+                                            // needs to to switch from one to
+                                            // another automatic fallback to full
+                                            // recovery to be applied
+
 static uint32_t    s_serial         = 0x01;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:18223]: Offtop: this static s_serial really could use some comment.

File: src/os/bluestore/BlueStore.cc [hunk 8]
```diff
@@ -18500,6 +18511,36 @@ int BlueStore::copy_allocator(Allocator* src_alloc, Allocator* dest_alloc, uint6
...
+
+//-----------------------------------------------------------------------------------
+int calc_allocator_image_trailer_size()
+{
+  utime_t                 timestamp       = ceph_clock_now();
+  uint64_t                extent_count    = -1;
+  uint64_t                allocation_size = -1;
+  uint32_t                crc             = -1;
+  bufferlist              trailer_bl;
+  allocator_image_trailer trailer(timestamp, s_format_version, s_serial, extent_count, allocation_size);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc]: dummy_format_version

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -18513,12 +18554,359 @@ static uint32_t flush_extent_buffer_with_crc(BlueFS::FileWriter *p_handle, const
...
+  {
+    bufferlist trailer_bl,temp_bl;
+    int        read_bytes = bluefs->read(p_handle, offset, trailer_size, &temp_bl, nullptr);
+    if (read_bytes != trailer_size) {
+      derr << "Failed bluefs->read() for trailer::read_bytes=" << read_bytes << ", req_bytes=" << trailer_size << dendl;
+      return -1;
+    }
+    offset += read_bytes;
+
+    trailer_bl.claim_append(temp_bl);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:18568]: Read directly into trailer_bl?

File: src/os/bluestore/BlueStore.cc [hunk 10]
```diff
@@ -18500,6 +18511,36 @@ int BlueStore::copy_allocator(Allocator* src_alloc, Allocator* dest_alloc, uint6
...
+  allocator_image_trailer trailer(timestamp, s_format_version, s_serial, extent_count, allocation_size);
+
+  encode(trailer, trailer_bl);
+  crc = trailer_bl.crc32c(crc);
+  encode(crc, trailer_bl);
+  return trailer_bl.length();
+}
+
 //-----------------------------------------------------------------------------------
 static uint32_t flush_extent_buffer_with_crc(BlueFS::FileWriter *p_handle, const char* buffer, const char *p_curr, uint32_t crc)
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:18480]: Lets make it "flush_buffer_with_crc". Half of the uses are just bits.

File: src/os/bluestore/BlueStore.cc [hunk 11]
```diff
@@ -18512,33 +18512,19 @@ int BlueStore::copy_allocator(Allocator* src_alloc, Allocator* dest_alloc, uint6
 }
 
 //-----------------------------------------------------------------------------------
-size_t calc_allocator_image_header_size()
+size_t calc_allocator_image_field_size(bool trailer)
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:18464]: The simplification of calc_allocator_image_header/tailer_size is very good, 
but switch to calc_allocator_image_field_size(bool trailer) makes it less clear.

**Review by aclamk — COMMENTED** [review]
Have you considered a method of mixed encoding?
Where high fragmentation exist put bitmap fragment, where large free region exists, put extent.

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:19376]: good catch

File: src/os/bluestore/BlueStore.cc [hunk 12]
```diff
@@ -18513,12 +18554,359 @@ static uint32_t flush_extent_buffer_with_crc(BlueFS::FileWriter *p_handle, const
...
+  ceph_assert(p_handle);
+  ceph_assert(header);
+  {
+    bufferlist trailer_bl,temp_bl;
+    int        read_bytes = bluefs->read(p_handle, offset, trailer_size, &temp_bl, nullptr);
+    if (read_bytes != trailer_size) {
+      derr << "Failed bluefs->read() for trailer::read_bytes=" << read_bytes << ", req_bytes=" << trailer_size << dendl;
+      return -1;
+    }
+    offset += read_bytes;
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:18566]: this is used in bluefs->read call above

File: src/os/bluestore/BlueStore.cc (hunk 9 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:18568]: BlueFS::_read invalidated output buffer prior to reading into it

File: src/os/bluestore/BlueStore.cc [hunk 13]
```diff
@@ -18516,11 +18516,12 @@ static uint32_t flush_extent_buffer_with_crc(BlueFS::FileWriter *p_handle, const
...
 // write the allocator to a flat bluefs file - 4K extents at a time
 //-----------------------------------------------------------------------------------
-int BlueStore::store_allocator(Allocator* src_allocator)
+int BlueStore::store_allocator(Allocator* src_allocator, const char* filename)
 {
+  dout(5) << __func__ << dendl;
   // when storing allocations to file we must be sure there is no background compactions
   // the easiest way to achieve it is to make sure db is closed
-  ceph_assert(db == nullptr);
+  ceph_assert(db == nullptr || src_allocator != alloc);
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:18846]: IIUC we don't care about db state if src_allocator differs from the primary bluestore allocator.

File: src/os/bluestore/BlueStore.cc [hunk 14]
```diff
@@ -18513,12 +18554,359 @@ static uint32_t flush_extent_buffer_with_crc(BlueFS::FileWriter *p_handle, const
...
+  if (bitmap == nullptr) {
+    derr << "Unable to save allocator in v2 format,"
+         << " bitmap representation is unsupported"
+         << dendl;
+    ret = -1;
+  } else {
+    const char* buffer = (const char*)bitmap;
+    uint64_t size = count * sizeof(*bitmap);
+    while (size > 0) {
+      size_t l = std::min(size, MAX_ENTRY_SIZE_V2);
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:18808]: switch to 64K chunks

File: src/os/bluestore/Allocator.h (hunk 3 above)
Comment (ifed01) [reply src/os/bluestore/Allocator.h:58]: IMO neither is ideal.  But I'm hesitant to expose BitmapAllocator to BlueStore. On the other hand if we do such an exposure then we can implement all the mechanics without dynamic_cast usage which is a bit better.

**Comment by ifed01** [comment]
> Have you considered a method of mixed encoding? Where high fragmentation exist put bitmap fragment, where large free region exists, put extent.

This looks to me like an overkill for this task - the approach like that might make sense for run-time data structures but for persistent data I don't see much benefit while the complexity is higher..

**Comment by ifed01** [comment]
unstale

**Comment by ifed01** [comment]
jenkins test api

**Comment by ifed01** [comment]
ceph test api

**Comment by ifed01** [comment]
unstale

**Comment by ifed01** [comment]
jenkins test api

**Comment by ifed01** [comment]
FYI:
commits: 
* fix BlueFS::foreach_block_extents (57daeaf07e39f808d34952f9c0f8a9afa44d64d6) 
* test/store_test: get rid off explicit offset specifications in shared (c915d18027d62f5cae8751f7aa743b215df3a115)

have standalone PRs https://github.com/ceph/ceph/pull/56985 and https://github.com/ceph/ceph/pull/55130 respectively.
Included here to get store tests running successfully..

**Comment by ifed01** [comment]
ceph api test

**Comment by ifed01** [comment]
jenkins api test

**Comment by ifed01** [comment]
jenkins test api

**Comment by ifed01** [comment]
jenkins test windows

**Comment by ifed01** [comment]
jenkins test windows

**Pull request closed without merging**